lib/scanner.js

```diff
@@ -60,6 +54,32 @@ class Scanner {
   rewind (pos) {
     this.pos = pos
   }
+
+  enter (type, content) {
+    const position = { start: this.position() }
+    return Array.isArray(content)
+      ? { type, children: content, position }
+      : { type, value: content, position }
+  }
+
+  exit (node) {
+    node.position.end = this.position()
+    return node
+  }
+
+  abort (node, expectedTokens) {
```

Comment:
Another idea: if this is too much for every `abort`, we can also add `fail` or something like that. We can move the error to that method, and only use `abort` for rewinding to the starting position of the node.

> Note that aborting failed parsing of a token, always rewinds the position back to start. Right now, that seems to work pretty good but I'm not sure how good that works with @bcoe's new recursive body parsing.

Reply:
I'm not convinced that calling `abort` would cause any issues, let's hold off on adding an additional method until we know we need to differentiate between an abort and failure.